documentation-website/Writerside/snippets/exposed-dao/src/main/kotlin/org/example/entities/StarWarsWFilmWithRankEntity.kt

```diff
@@ -0,0 +1,25 @@
+package org.example.entities
+
+import org.example.tables.StarWarsWFilmsWithRankTable
+import org.jetbrains.exposed.dao.IntEntity
+import org.jetbrains.exposed.dao.IntEntityClass
+import org.jetbrains.exposed.dao.id.EntityID
+import org.jetbrains.exposed.sql.Op
+import org.jetbrains.exposed.sql.Query
+
+class StarWarsWFilmWithRankEntity(id: EntityID<Int>) : IntEntity(id) {
+    var sequelId by StarWarsWFilmsWithRankTable.sequelId
+    var name by StarWarsWFilmsWithRankTable.name
+    var rating by StarWarsWFilmsWithRankTable.rating
+
+    val rank: Long
+        get() = readValues[StarWarsWFilmsWithRankTable.rank]
+
+    companion object : IntEntityClass<StarWarsWFilmWithRankEntity>(StarWarsWFilmsWithRankTable) {
+        override fun searchQuery(op: Op<Boolean>): Query {
+            return super.searchQuery(op).adjustSelect {
+                select(columns + StarWarsWFilmsWithRankTable.rank)
```

Comment:
@bog-walk In the original example line 21 has `.set` in the end. I removed it to pass the detekt check because otherwise it produced the following error:
`Type mismatch: inferred type is FieldSet but Query was expected`
As it is now, the `ReadExamples.readComputedField` example (currently commented out) results in this error:
```shell
Exception in thread "main" java.lang.IllegalStateException: RANK() OVER( ORDER BY STARWARSWFILMSWITHRANK.RATING DESC) is not in record set
        at org.jetbrains.exposed.sql.ResultRow.getExpressionIndex(ResultRow.kt:128)
        at org.jetbrains.exposed.sql.ResultRow.getRaw(ResultRow.kt:109)
        at org.jetbrains.exposed.sql.ResultRow.getInternal$lambda$5(ResultRow.kt:72)
        at org.jetbrains.exposed.sql.ResultRow$ResultRowCache.cached(ResultRow.kt:207)
        at org.jetbrains.exposed.sql.ResultRow.getInternal(ResultRow.kt:71)
        at org.jetbrains.exposed.sql.ResultRow.get(ResultRow.kt:40)
        at org.example.entities.StarWarsWFilmWithRankEntity.getRank(StarWarsWFilmWithRankEntity.kt:16)
        at org.example.examples.ReadExamples.readComputedField(ReadExamples.kt:117)
```

Can you advise how to fix it?

Reply:
The `entityCache.clear()` from the original example is what's needed to fix this (or perform each operation in a separate transaction).

When `new()` creates an entity, it's cached. Then using `find()` first checks the cache for the entity and, only if it isn't found, then pings the database to find the record and returns it as a new entity.
So the current setup means the entity is in the cache, and is returned by `find()`. But it throws that exception because the cached entity only stores the data provided to `new()`. The window function rank is a database generated value so we need to make sure the entity is being retrieved from the database instead of the cache.

So we have to manually clear the cache since we're running these samples in a single transaction:
```kt
fun readComputedField() {
    transaction {
        StarWarsWFilmWithRankEntity.new {
            sequelId = MOVIE_SEQUELID
            name = "The Last Jedi"
            rating = MOVIE_RATING
        }

        entityCache.clear()

        StarWarsWFilmWithRankEntity
            .find { StarWarsWFilmsWithRankTable.name like "The%" }
            .map { it.name to it.rank }
    }
}
```

The alternative would be to simulate what a user more likely does, which is create entities in 1 transaction, then perform queries in a separate transaction:
```kt
// the way App.main() is set up, this would require changing the project's database config
Database.connect(
    "jdbc:h2:mem:test",
    "org.h2.Driver",
    databaseConfig = DatabaseConfig { useNestedTransactions = true }
)

fun readComputedField() {
    transaction {
        StarWarsWFilmWithRankEntity.new {
            sequelId = MOVIE_SEQUELID
            name = "The Last Jedi"
            rating = MOVIE_RATING
        }
    }

    transaction {
        StarWarsWFilmWithRankEntity
            .find { StarWarsWFilmsWithRankTable.name like "The%" }
            .map { it.name to it.rank }
    }
}
```